### ISSUE TITLE
MueLu: self-reported time does not always match time reported by application for Hierarchy setup

### ISSUE THREAD

**Issue by jhux2** [trilinos/Trilinos#1848]
@trilinos/muelu 

In some cases, I've observed that MueLu's self-reported time for Hierarchy setup is significantly less than the time reported by the Nalu application.

Both use Teuchos timers.  Nalu puts a timer around the call to CreateTpetraPreconditioner,  MueLu starts/stops its timer at the beginning/end of Hierarchy::Setup.

Here's an example of the discrepancy:
```
MueLu: Hierarchy: Setup (total)          213.4 (600)          219.5 (721.9)        229.1 (1000)
nalu MueLu preconditioner setup          247.7 (200)          248.1 (200)          252.4 (200)
```

**Comment by aprokop** [comment]
@jhux2 Well, one thing seems to be that 
```c++
      std::string xmlFileName = config_->muelu_xml_file();
      mueluPreconditioner_ = MueLu::CreateTpetraPreconditioner<SC,LO,GO,NO>(Teuchos::RCP<Tpetra::Operator<SC,LO,GO,NO> >(matrix_), xmlFileName, coords_);
```
so you read the xml file from disc each time which may depend on the file system. I think it could be changed to actually store MueLu's parameters as a `ParameterList`.

**Comment by jhux2** [comment]
@aprokop Good point.  Also, the CreateTpetraPreconditioner interface itself is completely untimed.

**Comment by cgcgcg** [comment]
@jhux2 @lucbv Is this still an issue, or can we close this one?

**Comment by lucbv** [comment]
I think we can unless we think it is important to add a timer around the call to `Teuchos::updateParametersFromXmlFileAndBroadcast(...)` which might not be the worst idea in the world considering it is highly IO and MPI dependent?
What do you think @cgcgcg

**Comment by cgcgcg** [comment]
So there is still a big gap between the Nalu timer and the MueLu one?

**Comment by lucbv** [comment]
I am not sure how big it is but I am sure that we are not timing the xml read-out phase, likely because it is not obvious how to nest the timers correctly when that adapter calls the parameter list interface directly.

**Comment by cgcgcg** [comment]
If it's really the call to `Teuchos::updateParametersFromXmlFileAndBroadcast`, then Nalu should just read the xml once and pass a ParameterList instead. If it's something else, it'd be good to know.

**Comment by jhux2** [comment]
Afaik, Nalu only reads the XML once.

**Comment by jjellio** [comment]
I don't know if it's related, but a while back we had a similar concern w/EMPIRE.  I believe Paul Lin and I determined it was due to a lack of Barrier after MueLu reports.  I'm not sure where exactly we tracked that (probably in that single gitlab tracker we use for all ATDM apps ;)  )
